### PR TITLE
Enhance nav2 dwb controller

### PR DESCRIPTION
**Pull request by DylanDeCoeyer-Quimesis**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   |  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on |  |

---

## Description of contribution in a few bullet points

* Add a dedicated distance threshold for the path shortening mechanism (enabled w/ "shorten_transformed_plan").
* Enhance path pruning and shortening algorithm.
* Add a parameter to define a maximum deviation from the path.

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by SteveMacenski** [comment]
https://app.circleci.com/pipelines/github/ros-planning/navigation2/8664/workflows/0e23fddf-71eb-44e4-9afd-2c7e24c3467d/jobs/28715

Check CI output for a number of linting errors

File: nav2_util/include/nav2_util/geometry_utils.hpp [hunk 1]
```diff
@@ -144,16 +144,17 @@ inline Iter min_by(Iter begin, Iter end, Getter getCompareVal)
  * Find first element in iterator that is greater integrated distance than comparevalue
  */
 template<typename Iter, typename Getter>
-inline Iter first_after_integrated_distance(Iter begin, Iter end, Getter getCompareVal)
+inline Iter first_after_integrated_distance(Iter begin, Iter end, Getter getCompareVal, bool backward=false)
```
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/geometry_utils.hpp:147]: I believe by the nature of this function being templated on iterators, you could instead give it reverser iterators and require no changes.

File: nav2_dwb_controller/dwb_core/src/dwb_local_planner.cpp [hunk 2]
```diff
@@ -474,51 +482,60 @@ DWBLocalPlanner::transformGlobalPlan(
...
+  {
+    double distance = euclidean_distance(robot_pose.pose, *it);
+    if (distance <= smallest_distance)
+    {
+      smallest_distance = distance;
+      closest_pose = it;
+    }
+    else
+    {
+      // as soon as the distance increases, stop searching
```
Comment (SteveMacenski) [inline nav2_dwb_controller/dwb_core/src/dwb_local_planner.cpp:498]: This is untrue for an arbitrary path

File: nav2_util/include/nav2_util/geometry_utils.hpp (hunk 1 above)
Comment (DylanDeCoeyer-Quimesis) [reply nav2_util/include/nav2_util/geometry_utils.hpp]: Right, I updated the commit accordingly.

File: nav2_dwb_controller/dwb_core/src/dwb_local_planner.cpp (hunk 2 above)
Comment (DylanDeCoeyer-Quimesis) [reply nav2_dwb_controller/dwb_core/src/dwb_local_planner.cpp]: Indeed, I made a new suggestion.

File: nav2_dwb_controller/dwb_core/src/dwb_local_planner.cpp [hunk 3]
```diff
@@ -469,57 +477,62 @@ DWBLocalPlanner::transformGlobalPlan(
...
-  double transform_end_threshold;
-  if (shorten_transformed_plan_) {
-    transform_end_threshold = std::min(dist_threshold, prune_dist);
-  } else {
-    transform_end_threshold = dist_threshold;
-  }
-
-  // Find the first pose in the global plan that's further than prune distance
-  // from the robot using integrated distance
-  auto prune_point = nav2_util::geometry_utils::first_after_integrated_distance(
```
Comment (SteveMacenski) [inline nav2_dwb_controller/dwb_core/src/dwb_local_planner.cpp:504]: This is to help with paths with overlapping segments to select the closest one to the actual robot. WIthout this, you're not guaranteed to pick the first of 2 similarly positioned points on a "crossing" of the path -- instead relying on a coin flip about which is "absolutely" closer by small floating point error.

Comment (DylanDeCoeyer-Quimesis) [reply nav2_dwb_controller/dwb_core/src/dwb_local_planner.cpp:504]: Hmm this is a complexe issue. For now, I don't see any way to find with certitude the closest point to the robot. We could bound the research range using the closest point found at the previous iteration, and the maximum speed provided as parameter (max_vel_x/y) multiplied by the controller frequency. Still we cannot exclude that the robot moves faster than this.

**Comment by DylanDeCoeyer-Quimesis** [comment]
I created a new PR to merge the first commit separately: https://github.com/ros-planning/navigation2/pull/3374

**Comment by SteveMacenski** [comment]
#3374 supersedes

**Pull request closed without merging**